idaes/apps/caprese/dynamic_builder.py

```diff
@@ -0,0 +1,166 @@
+# Research Corporation, et al.  All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and
+# license information.
+#################################################################################
+"""
+Class for performing plant, MHE, NMPC simulations of IDAES flowsheets.
+"""
+
+import random
+from pyomo.environ import (
+        ComponentUID,
+        )
+from pyomo.util.slices import slice_component_along_sets
+from pyomo.common.config import ConfigDict, ConfigValue
+
+from idaes.apps.caprese.dynamic_block import DynamicBlock
+from idaes.apps.caprese.controller import ControllerBlock
+from idaes.apps.caprese.estimator import EstimatorBlock
+from idaes.apps.caprese.common.config import VariableCategory as VC
+
+__author__ = "Robert Parker, David Thierry, and Kuan-Han Lin"
+
+
+def use_CUID_to_capture_vars_t0_in_given_model(cuid_list, 
+                                               target_t0,
+                                               target_mod):
+    varlist_in_target_model = [cuid.find_component_on(target_mod)[target_t0]
+                               for cuid in cuid_list]
+            
+    return varlist_in_target_model
+            
+class DynamicSim(object):
+    def __init__(self,
+                 plant_model = None,
+                 plant_time_set = None,
+                 estimator_model = None,
+                 estimator_time_set = None,
+                 controller_model = None, 
+                 controller_time_set = None,
+                 inputs_at_t0 = None,
+                 measurements_at_t0 = None,
+                 sample_time = None,):
+        
+        self.input_cuids = [
+                            ComponentUID(
+                            slice_component_along_sets(comp, (plant_time_set,)))
+                            for comp in inputs_at_t0
+                            ]
+        self.measurement_cuids = [
+                                    ComponentUID(
+                                    slice_component_along_sets(comp, (plant_time_set,)))
+                                    for comp in measurements_at_t0
+                                    ]
+        
+        self.plant_is_existing = False
+        self.estimator_is_existing = False
+        self.controller_is_existing = False
+        
+        #----------------------------------------------------------------------                
+        if plant_model:
+            self.plant_is_existing = True
+            
+            # Capture vars in plant model
+            plant_inputs_t0 = inputs_at_t0
+            plant_measurements_t0 = measurements_at_t0
+            
+            self.plant = DynamicBlock(
+                    model=plant_model,
+                    time=plant_time_set,
+                    inputs=plant_inputs_t0,
+                    measurements=plant_measurements_t0,
+                    )
+            self.plant.construct()
+            
```

Comment:
Here and elsewhere, whitespace-only lines shouldn't have any characters (spaces/tabs) in them. Not something you need to go and fix, but just FYI for future code.